### ISSUE TITLE
--ignore is broken

### ISSUE THREAD

**Issue by mjackson** [babel/babel#1539]
I can't get the `--ignore` switch to ignore my test files, no matter which glob pattern I try. Here are the results from trying `--ignore *-test.js`, `--ignore modules/__tests__/*-test.js`, and `--ignore modules/**/*-test.js`:

```
michael:~/Projects/react-router$ babel ./modules -d lib --optional es7.classProperties --ignore *-test.js
modules/AbstractHistory.js -> lib/AbstractHistory.js
modules/AsyncUtils.js -> lib/AsyncUtils.js
modules/BrowserHistory.js -> lib/BrowserHistory.js
modules/DOMHistory.js -> lib/DOMHistory.js
modules/DOMUtils.js -> lib/DOMUtils.js
modules/HashHistory.js -> lib/HashHistory.js
modules/History.js -> lib/History.js
modules/Link.js -> lib/Link.js
modules/Location.js -> lib/Location.js
modules/NativeHistory.js -> lib/NativeHistory.js
modules/Navigation.js -> lib/Navigation.js
modules/NavigationMixin.js -> lib/NavigationMixin.js
modules/NavigationTypes.js -> lib/NavigationTypes.js
modules/PathUtils.js -> lib/PathUtils.js
modules/PropTypes.js -> lib/PropTypes.js
modules/RefreshHistory.js -> lib/RefreshHistory.js
modules/Route.js -> lib/Route.js
modules/ScrollHistory.js -> lib/ScrollHistory.js
modules/State.js -> lib/State.js
modules/StateMixin.js -> lib/StateMixin.js
modules/TestUtils.js -> lib/TestUtils.js
modules/Transition.js -> lib/Transition.js
modules/TransitionMixin.js -> lib/TransitionMixin.js
modules/__tests__/BrowserHistory-test.js -> lib/__tests__/BrowserHistory-test.js
modules/__tests__/HashHistory-test.js -> lib/__tests__/HashHistory-test.js
modules/__tests__/History-test.js -> lib/__tests__/History-test.js
modules/__tests__/Link-test.js -> lib/__tests__/Link-test.js
modules/__tests__/Location-test.js -> lib/__tests__/Location-test.js
modules/__tests__/PathUtils-test.js -> lib/__tests__/PathUtils-test.js
modules/__tests__/RefreshHistory-test.js -> lib/__tests__/RefreshHistory-test.js
modules/__tests__/Route-test.js -> lib/__tests__/Route-test.js
modules/__tests__/TransitionMixin-test.js -> lib/__tests__/TransitionMixin-test.js
modules/__tests__/createRouter-test.js -> lib/__tests__/createRouter-test.js
modules/__tests__/createRoutesFromReactChildren-test.js -> lib/__tests__/createRoutesFromReactChildren-test.js
modules/__tests__/describeHistory.js -> lib/__tests__/describeHistory.js
modules/__tests__/findMatch-test.js -> lib/__tests__/findMatch-test.js
modules/behaviors/ImitateBrowserBehavior.js -> lib/behaviors/ImitateBrowserBehavior.js
modules/behaviors/ScrollToTopBehavior.js -> lib/behaviors/ScrollToTopBehavior.js
modules/createRouter.js -> lib/createRouter.js
modules/createRoutesFromReactChildren.js -> lib/createRoutesFromReactChildren.js
modules/findMatch.js -> lib/findMatch.js
modules/index.js -> lib/index.js
modules/isReactChildren.js -> lib/isReactChildren.js


michael:~/Projects/react-router$ babel ./modules -d lib --optional es7.classProperties --ignore modules/__tests__/*-test.js
modules/AbstractHistory.js -> lib/AbstractHistory.js
modules/AsyncUtils.js -> lib/AsyncUtils.js
modules/BrowserHistory.js -> lib/BrowserHistory.js
modules/DOMHistory.js -> lib/DOMHistory.js
modules/DOMUtils.js -> lib/DOMUtils.js
modules/HashHistory.js -> lib/HashHistory.js
modules/History.js -> lib/History.js
modules/Link.js -> lib/Link.js
modules/Location.js -> lib/Location.js
modules/NativeHistory.js -> lib/NativeHistory.js
modules/Navigation.js -> lib/Navigation.js
modules/NavigationMixin.js -> lib/NavigationMixin.js
modules/NavigationTypes.js -> lib/NavigationTypes.js
modules/PathUtils.js -> lib/PathUtils.js
modules/PropTypes.js -> lib/PropTypes.js
modules/RefreshHistory.js -> lib/RefreshHistory.js
modules/Route.js -> lib/Route.js
modules/ScrollHistory.js -> lib/ScrollHistory.js
modules/State.js -> lib/State.js
modules/StateMixin.js -> lib/StateMixin.js
modules/TestUtils.js -> lib/TestUtils.js
modules/Transition.js -> lib/Transition.js
modules/TransitionMixin.js -> lib/TransitionMixin.js
modules/__tests__/HashHistory-test.js -> lib/__tests__/HashHistory-test.js
modules/__tests__/History-test.js -> lib/__tests__/History-test.js
modules/__tests__/Link-test.js -> lib/__tests__/Link-test.js
modules/__tests__/Location-test.js -> lib/__tests__/Location-test.js
modules/__tests__/PathUtils-test.js -> lib/__tests__/PathUtils-test.js
modules/__tests__/RefreshHistory-test.js -> lib/__tests__/RefreshHistory-test.js
modules/__tests__/Route-test.js -> lib/__tests__/Route-test.js
modules/__tests__/TransitionMixin-test.js -> lib/__tests__/TransitionMixin-test.js
modules/__tests__/createRouter-test.js -> lib/__tests__/createRouter-test.js
modules/__tests__/createRoutesFromReactChildren-test.js -> lib/__tests__/createRoutesFromReactChildren-test.js
modules/__tests__/describeHistory.js -> lib/__tests__/describeHistory.js
modules/__tests__/findMatch-test.js -> lib/__tests__/findMatch-test.js
modules/behaviors/ImitateBrowserBehavior.js -> lib/behaviors/ImitateBrowserBehavior.js
modules/behaviors/ScrollToTopBehavior.js -> lib/behaviors/ScrollToTopBehavior.js
modules/createRouter.js -> lib/createRouter.js
modules/createRoutesFromReactChildren.js -> lib/createRoutesFromReactChildren.js
modules/findMatch.js -> lib/findMatch.js
modules/index.js -> lib/index.js
modules/isReactChildren.js -> lib/isReactChildren.js
modules/__tests__/HashHistory-test.js -> lib/modules/__tests__/HashHistory-test.js
modules/__tests__/History-test.js -> lib/modules/__tests__/History-test.js
modules/__tests__/Link-test.js -> lib/modules/__tests__/Link-test.js
modules/__tests__/Location-test.js -> lib/modules/__tests__/Location-test.js
modules/__tests__/PathUtils-test.js -> lib/modules/__tests__/PathUtils-test.js
modules/__tests__/RefreshHistory-test.js -> lib/modules/__tests__/RefreshHistory-test.js
modules/__tests__/Route-test.js -> lib/modules/__tests__/Route-test.js
modules/__tests__/TransitionMixin-test.js -> lib/modules/__tests__/TransitionMixin-test.js
modules/__tests__/createRouter-test.js -> lib/modules/__tests__/createRouter-test.js
modules/__tests__/createRoutesFromReactChildren-test.js -> lib/modules/__tests__/createRoutesFromReactChildren-test.js
modules/__tests__/findMatch-test.js -> lib/modules/__tests__/findMatch-test.js


michael:~/Projects/react-router$ babel ./modules -d lib --optional es7.classProperties --ignore modules/**/*-test.js
modules/AbstractHistory.js -> lib/AbstractHistory.js
modules/AsyncUtils.js -> lib/AsyncUtils.js
modules/BrowserHistory.js -> lib/BrowserHistory.js
modules/DOMHistory.js -> lib/DOMHistory.js
modules/DOMUtils.js -> lib/DOMUtils.js
modules/HashHistory.js -> lib/HashHistory.js
modules/History.js -> lib/History.js
modules/Link.js -> lib/Link.js
modules/Location.js -> lib/Location.js
modules/NativeHistory.js -> lib/NativeHistory.js
modules/Navigation.js -> lib/Navigation.js
modules/NavigationMixin.js -> lib/NavigationMixin.js
modules/NavigationTypes.js -> lib/NavigationTypes.js
modules/PathUtils.js -> lib/PathUtils.js
modules/PropTypes.js -> lib/PropTypes.js
modules/RefreshHistory.js -> lib/RefreshHistory.js
modules/Route.js -> lib/Route.js
modules/ScrollHistory.js -> lib/ScrollHistory.js
modules/State.js -> lib/State.js
modules/StateMixin.js -> lib/StateMixin.js
modules/TestUtils.js -> lib/TestUtils.js
modules/Transition.js -> lib/Transition.js
modules/TransitionMixin.js -> lib/TransitionMixin.js
modules/__tests__/HashHistory-test.js -> lib/__tests__/HashHistory-test.js
modules/__tests__/History-test.js -> lib/__tests__/History-test.js
modules/__tests__/Link-test.js -> lib/__tests__/Link-test.js
modules/__tests__/Location-test.js -> lib/__tests__/Location-test.js
modules/__tests__/PathUtils-test.js -> lib/__tests__/PathUtils-test.js
modules/__tests__/RefreshHistory-test.js -> lib/__tests__/RefreshHistory-test.js
modules/__tests__/Route-test.js -> lib/__tests__/Route-test.js
modules/__tests__/TransitionMixin-test.js -> lib/__tests__/TransitionMixin-test.js
modules/__tests__/createRouter-test.js -> lib/__tests__/createRouter-test.js
modules/__tests__/createRoutesFromReactChildren-test.js -> lib/__tests__/createRoutesFromReactChildren-test.js
modules/__tests__/describeHistory.js -> lib/__tests__/describeHistory.js
modules/__tests__/findMatch-test.js -> lib/__tests__/findMatch-test.js
modules/behaviors/ImitateBrowserBehavior.js -> lib/behaviors/ImitateBrowserBehavior.js
modules/behaviors/ScrollToTopBehavior.js -> lib/behaviors/ScrollToTopBehavior.js
modules/createRouter.js -> lib/createRouter.js
modules/createRoutesFromReactChildren.js -> lib/createRoutesFromReactChildren.js
modules/findMatch.js -> lib/findMatch.js
modules/index.js -> lib/index.js
modules/isReactChildren.js -> lib/isReactChildren.js
modules/__tests__/HashHistory-test.js -> lib/modules/__tests__/HashHistory-test.js
modules/__tests__/History-test.js -> lib/modules/__tests__/History-test.js
modules/__tests__/Link-test.js -> lib/modules/__tests__/Link-test.js
modules/__tests__/Location-test.js -> lib/modules/__tests__/Location-test.js
modules/__tests__/PathUtils-test.js -> lib/modules/__tests__/PathUtils-test.js
modules/__tests__/RefreshHistory-test.js -> lib/modules/__tests__/RefreshHistory-test.js
modules/__tests__/Route-test.js -> lib/modules/__tests__/Route-test.js
modules/__tests__/TransitionMixin-test.js -> lib/modules/__tests__/TransitionMixin-test.js
modules/__tests__/createRouter-test.js -> lib/modules/__tests__/createRouter-test.js
modules/__tests__/createRoutesFromReactChildren-test.js -> lib/modules/__tests__/createRoutesFromReactChildren-test.js
modules/__tests__/findMatch-test.js -> lib/modules/__tests__/findMatch-test.js
```

What am I doing wrong?

Note: when using either `modules/__tests__/*-test.js` or `modules/**/*-test.js` the files are actually compiled **twice** instead of being ignored.

**Comment by sebmck** [comment]
Try `--ignore **/__tests__/*` or `--ignore ./modules/__tests__/*`.

**Comment by mjackson** [comment]
@sebmck Still no joy :(

`**/__tests__/*` actually makes the problem worse, running files that are already in the output dir back through the compiler. `./modules/__tests__/*` behaves the same as the version without the leading dot slash.

```
michael:~/Projects/react-router$ babel ./modules -d lib --optional es7.classProperties --ignore **/__tests__/*
modules/AbstractHistory.js -> lib/AbstractHistory.js
modules/AsyncUtils.js -> lib/AsyncUtils.js
modules/BrowserHistory.js -> lib/BrowserHistory.js
modules/DOMHistory.js -> lib/DOMHistory.js
modules/DOMUtils.js -> lib/DOMUtils.js
modules/HashHistory.js -> lib/HashHistory.js
modules/History.js -> lib/History.js
modules/Link.js -> lib/Link.js
modules/Location.js -> lib/Location.js
modules/NativeHistory.js -> lib/NativeHistory.js
modules/Navigation.js -> lib/Navigation.js
modules/NavigationMixin.js -> lib/NavigationMixin.js
modules/NavigationTypes.js -> lib/NavigationTypes.js
modules/PathUtils.js -> lib/PathUtils.js
modules/PropTypes.js -> lib/PropTypes.js
modules/RefreshHistory.js -> lib/RefreshHistory.js
modules/Route.js -> lib/Route.js
modules/ScrollHistory.js -> lib/ScrollHistory.js
modules/State.js -> lib/State.js
modules/StateMixin.js -> lib/StateMixin.js
modules/TestUtils.js -> lib/TestUtils.js
modules/Transition.js -> lib/Transition.js
modules/TransitionMixin.js -> lib/TransitionMixin.js
modules/__tests__/BrowserHistory-test.js -> lib/__tests__/BrowserHistory-test.js
modules/__tests__/HashHistory-test.js -> lib/__tests__/HashHistory-test.js
modules/__tests__/History-test.js -> lib/__tests__/History-test.js
modules/__tests__/Link-test.js -> lib/__tests__/Link-test.js
modules/__tests__/Location-test.js -> lib/__tests__/Location-test.js
modules/__tests__/PathUtils-test.js -> lib/__tests__/PathUtils-test.js
modules/__tests__/RefreshHistory-test.js -> lib/__tests__/RefreshHistory-test.js
modules/__tests__/Route-test.js -> lib/__tests__/Route-test.js
modules/__tests__/TransitionMixin-test.js -> lib/__tests__/TransitionMixin-test.js
modules/__tests__/createRouter-test.js -> lib/__tests__/createRouter-test.js
modules/__tests__/createRoutesFromReactChildren-test.js -> lib/__tests__/createRoutesFromReactChildren-test.js
modules/__tests__/describeHistory.js -> lib/__tests__/describeHistory.js
modules/__tests__/findMatch-test.js -> lib/__tests__/findMatch-test.js
modules/behaviors/ImitateBrowserBehavior.js -> lib/behaviors/ImitateBrowserBehavior.js
modules/behaviors/ScrollToTopBehavior.js -> lib/behaviors/ScrollToTopBehavior.js
modules/createRouter.js -> lib/createRouter.js
modules/createRoutesFromReactChildren.js -> lib/createRoutesFromReactChildren.js
modules/findMatch.js -> lib/findMatch.js
modules/index.js -> lib/index.js
modules/isReactChildren.js -> lib/isReactChildren.js
lib/__tests__/HashHistory-test.js -> lib/lib/__tests__/HashHistory-test.js
lib/__tests__/History-test.js -> lib/lib/__tests__/History-test.js
lib/__tests__/Link-test.js -> lib/lib/__tests__/Link-test.js
lib/__tests__/Location-test.js -> lib/lib/__tests__/Location-test.js
lib/__tests__/PathUtils-test.js -> lib/lib/__tests__/PathUtils-test.js
lib/__tests__/RefreshHistory-test.js -> lib/lib/__tests__/RefreshHistory-test.js
lib/__tests__/Route-test.js -> lib/lib/__tests__/Route-test.js
lib/__tests__/TransitionMixin-test.js -> lib/lib/__tests__/TransitionMixin-test.js
lib/__tests__/createRouter-test.js -> lib/lib/__tests__/createRouter-test.js
lib/__tests__/createRoutesFromReactChildren-test.js -> lib/lib/__tests__/createRoutesFromReactChildren-test.js
lib/__tests__/describeHistory.js -> lib/lib/__tests__/describeHistory.js
lib/__tests__/findMatch-test.js -> lib/lib/__tests__/findMatch-test.js
modules/__tests__/BrowserHistory-test.js -> lib/modules/__tests__/BrowserHistory-test.js
modules/__tests__/HashHistory-test.js -> lib/modules/__tests__/HashHistory-test.js
modules/__tests__/History-test.js -> lib/modules/__tests__/History-test.js
modules/__tests__/Link-test.js -> lib/modules/__tests__/Link-test.js
modules/__tests__/Location-test.js -> lib/modules/__tests__/Location-test.js
modules/__tests__/PathUtils-test.js -> lib/modules/__tests__/PathUtils-test.js
modules/__tests__/RefreshHistory-test.js -> lib/modules/__tests__/RefreshHistory-test.js
modules/__tests__/Route-test.js -> lib/modules/__tests__/Route-test.js
modules/__tests__/TransitionMixin-test.js -> lib/modules/__tests__/TransitionMixin-test.js
modules/__tests__/createRouter-test.js -> lib/modules/__tests__/createRouter-test.js
modules/__tests__/createRoutesFromReactChildren-test.js -> lib/modules/__tests__/createRoutesFromReactChildren-test.js
modules/__tests__/describeHistory.js -> lib/modules/__tests__/describeHistory.js
modules/__tests__/findMatch-test.js -> lib/modules/__tests__/findMatch-test.js


michael:~/Projects/react-router$ babel ./modules -d lib --optional es7.classProperties --ignore ./modules/__tests__/*
modules/AbstractHistory.js -> lib/AbstractHistory.js
modules/AsyncUtils.js -> lib/AsyncUtils.js
modules/BrowserHistory.js -> lib/BrowserHistory.js
modules/DOMHistory.js -> lib/DOMHistory.js
modules/DOMUtils.js -> lib/DOMUtils.js
modules/HashHistory.js -> lib/HashHistory.js
modules/History.js -> lib/History.js
modules/Link.js -> lib/Link.js
modules/Location.js -> lib/Location.js
modules/NativeHistory.js -> lib/NativeHistory.js
modules/Navigation.js -> lib/Navigation.js
modules/NavigationMixin.js -> lib/NavigationMixin.js
modules/NavigationTypes.js -> lib/NavigationTypes.js
modules/PathUtils.js -> lib/PathUtils.js
modules/PropTypes.js -> lib/PropTypes.js
modules/RefreshHistory.js -> lib/RefreshHistory.js
modules/Route.js -> lib/Route.js
modules/ScrollHistory.js -> lib/ScrollHistory.js
modules/State.js -> lib/State.js
modules/StateMixin.js -> lib/StateMixin.js
modules/TestUtils.js -> lib/TestUtils.js
modules/Transition.js -> lib/Transition.js
modules/TransitionMixin.js -> lib/TransitionMixin.js
modules/__tests__/BrowserHistory-test.js -> lib/__tests__/BrowserHistory-test.js
modules/__tests__/HashHistory-test.js -> lib/__tests__/HashHistory-test.js
modules/__tests__/History-test.js -> lib/__tests__/History-test.js
modules/__tests__/Link-test.js -> lib/__tests__/Link-test.js
modules/__tests__/Location-test.js -> lib/__tests__/Location-test.js
modules/__tests__/PathUtils-test.js -> lib/__tests__/PathUtils-test.js
modules/__tests__/RefreshHistory-test.js -> lib/__tests__/RefreshHistory-test.js
modules/__tests__/Route-test.js -> lib/__tests__/Route-test.js
modules/__tests__/TransitionMixin-test.js -> lib/__tests__/TransitionMixin-test.js
modules/__tests__/createRouter-test.js -> lib/__tests__/createRouter-test.js
modules/__tests__/createRoutesFromReactChildren-test.js -> lib/__tests__/createRoutesFromReactChildren-test.js
modules/__tests__/describeHistory.js -> lib/__tests__/describeHistory.js
modules/__tests__/findMatch-test.js -> lib/__tests__/findMatch-test.js
modules/behaviors/ImitateBrowserBehavior.js -> lib/behaviors/ImitateBrowserBehavior.js
modules/behaviors/ScrollToTopBehavior.js -> lib/behaviors/ScrollToTopBehavior.js
modules/createRouter.js -> lib/createRouter.js
modules/createRoutesFromReactChildren.js -> lib/createRoutesFromReactChildren.js
modules/findMatch.js -> lib/findMatch.js
modules/index.js -> lib/index.js
modules/isReactChildren.js -> lib/isReactChildren.js
./modules/__tests__/HashHistory-test.js -> lib/modules/__tests__/HashHistory-test.js
./modules/__tests__/History-test.js -> lib/modules/__tests__/History-test.js
./modules/__tests__/Link-test.js -> lib/modules/__tests__/Link-test.js
./modules/__tests__/Location-test.js -> lib/modules/__tests__/Location-test.js
./modules/__tests__/PathUtils-test.js -> lib/modules/__tests__/PathUtils-test.js
./modules/__tests__/RefreshHistory-test.js -> lib/modules/__tests__/RefreshHistory-test.js
./modules/__tests__/Route-test.js -> lib/modules/__tests__/Route-test.js
./modules/__tests__/TransitionMixin-test.js -> lib/modules/__tests__/TransitionMixin-test.js
./modules/__tests__/createRouter-test.js -> lib/modules/__tests__/createRouter-test.js
./modules/__tests__/createRoutesFromReactChildren-test.js -> lib/modules/__tests__/createRoutesFromReactChildren-test.js
./modules/__tests__/describeHistory.js -> lib/modules/__tests__/describeHistory.js
./modules/__tests__/findMatch-test.js -> lib/modules/__tests__/findMatch-test.js
```

**Comment by sebmck** [comment]
So I've heavily improved/clean up the `shouldIgnore` algorithm. It'll crawl the entire path and match each part against your `ignore` filters as well as the entire part leading up to it. I'll push out a release shortly which will allow you to do:

```javascript
$ babel --ignore test-*.js
$ babel --ignore __tests__
$ babel --ignore "__tests__/*"
```

etc.

Also important to note that you should wrap your patterns in quotes so your shell doesn't explode them.

**Comment by sebmck** [comment]
Fixed/improved(?) as of 5.4.1, thanks!

**Comment by mjackson** [comment]
@sebmck thank **you**! works perfectly.
